File: src/main/java/com/nwmqpa/graphcall/nodes/Node.java
package com.nwmqpa.graphcall.nodes;

import com.nwmqpa.graphcall.interfaces.ICompilable;

import java.util.ArrayList;
import java.util.List;

public class Node implements ICompilable {

    private final String name;
    private final ICompilable parent;
    private final String nesting;

    private List<ICompilable> nodes;
    private List<ICompilable> parametersField;

    public Node(String name) {
        this(name, null, "");
    }

    private Node(String name, ICompilable parent, String nesting) {
        this.name = name;
        this.parent = parent;
        this.nesting = nesting;
        this.nodes = null;
        this.parametersField = null;
    }

    @Override
    public String compile() {
        StringBuilder sb = new StringBuilder();
        sb.append(nesting).append(this.name);
        if (this.parametersField != null) {
            sb.append("(");
            for (ICompilable parameterField: parametersField) {
                sb.append(parameterField.compile());
                if (parametersField.lastIndexOf(parameterField) != parametersField.size() - 1)
                    sb.append(", ");
            }
            sb.append(")");
        }
        if (this.nodes != null) {
            sb.append(" {\n");
            for (ICompilable node : nodes) {
                sb.append(node.compile());
            }
            sb.append(nesting + "}\n");

        }
        return sb.toString();
    }

    public Node addNode(String name) {
        if (this.nodes == null)
            this.nodes = new ArrayList<>();
        Node node = new Node(name, this, this.nesting + "\t");
        this.nodes.add(node);
        return (node);
    }

    public Node addEdge(String name) {
        if (this.nodes == null)
            this.nodes = new ArrayList<>();
        Edge edge = new Edge(name, this.nesting + "\t");
        this.nodes.add(edge);
        return this;
    }

    public Node buildNode() {
        return (Node) this.parent;
    }

    public Node addParameterField(String parameter, String value) {
        if (this.parametersField == null)
            this.parametersField = new ArrayList<>();
        ParameterField parameterField = new ParameterField(parameter, value);
        this.parametersField.add(parameterField);
        return this;
    }
}

File: python/graphcall/nodes/node.py
"""Module defining what a node is."""
from graphcall.interfaces.compilable import Compilable
from graphcall.nodes.edge import Edge
from graphcall.nodes.parameter_field import ParameterField


class Node(Compilable):
    """Node class."""

    def __init__(self, name: str):
        """Initialize the node."""
        self._init(name, None, "")

    def _init(self, name: str, parent: Compilable, nesting: str):
        """Initialize fully the node."""
        self.name = name
        self.parent = parent
        self.nesting = nesting
        self.nodes = None
        self.parametersField = None

    def addNode(self, name: str):
        """Add a node to the current node."""
        if self.nodes is None:
            self.nodes = []
        node = Node("")
        node._init(name, self, self.nesting + "\t")
        self.nodes.append(
            node
        )
        return node

    def compile(self) -> str:
        """Compile the current node into string"""
        result = self.nesting + self.name
        if self.parametersField is not None:
            result += "("
            for index in range(len(self.parametersField)):
                result += self.parametersField[index].compile()
                if index != len(self.parametersField) - 1:
                    result += ", "
            result += ")"
        if self.nodes is not None:
            result += " {\n"
            for index in range(len(self.nodes)):
                result += self.nodes[index].compile()
            result += self.nesting + "}\n"
        return result


    def buildNode(self):
        """Mark the current node as built."""
        return self.parent if self.parent is not None else self

    def addParameterField(self, parameter: str, value):
        """Add a parameter field to the current node."""
        if self.parametersField is None:
            self.parametersField = []
        self.parametersField.append(
            ParameterField(parameter, value)
        )
        return self

    def addEdge(self, name: str):
        if self.nodes is None:
            self.nodes = []
        self.nodes.append(
            Edge(name, self.nesting + "\t")
        )
        return self

    def __str__(self):
        """Return a str version of the query."""
        return self.compile()

    def __repl__(self):
        return str(self)
